flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java

```diff
@@ -0,0 +1,586 @@
+                        sinkFunction);
+        builder.setFailureHandler(new DummyRetryFailureHandler());
+
+        final OpensearchSink<String> sink = builder.build();
+        sink.setRuntimeContext(new MockStreamingRuntimeContext(false, 1, 0));
+        sink.open(new Configuration());
+        sink.close();
+
+        assertThat(sinkFunction.openCalled).isTrue();
+        assertThat(sinkFunction.closeCalled).isTrue();
+    }
+
+    private static class SimpleSinkFunction<String> implements OpensearchSinkFunction<String> {
+        private static final long serialVersionUID = -176739293659135148L;
+
+        @Override
+        public void process(String element, RuntimeContext ctx, RequestIndexer indexer) {
+            Map<java.lang.String, Object> json = new HashMap<>();
+            json.put("data", element);
+
+            indexer.add(Requests.indexRequest().index("index").type("type").id("id").source(json));
+        }
+    }
+
+    private static class SimpleClosableSinkFunction<String>
+            implements OpensearchSinkFunction<String> {
+
+        private static final long serialVersionUID = 1872065917794006848L;
+
+        private boolean openCalled;
+        private boolean closeCalled;
+
+        @Override
+        public void open() {
+            openCalled = true;
+        }
+
+        @Override
+        public void close() {
+            closeCalled = true;
+        }
+
+        @Override
+        public void process(String element, RuntimeContext ctx, RequestIndexer indexer) {}
+    }
+
+    private static class DummyRetryFailureHandler implements ActionRequestFailureHandler {
+        private static final long serialVersionUID = 5400023700099200745L;
+
+        @Override
+        public void onFailure(
+                ActionRequest action, Throwable failure, int restStatusCode, RequestIndexer indexer)
+                throws Throwable {
+            indexer.add(action);
+        }
+    }
+
+    private static Consumer<HttpResponse> createResponse(BulkItemResponse item) {
+        return response -> {
+            try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+                response.setStatusCode(200);
+                try (XContentBuilder builder =
+                        new XContentBuilder(JsonXContent.jsonXContent, baos)) {
+                    final BulkResponse bulkResponse =
+                            new BulkResponse(new BulkItemResponse[] {item}, 200);
+                    bulkResponse.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                }
+                response.setEntity(
+                        new ByteArrayEntity(baos.toByteArray(), ContentType.APPLICATION_JSON));
+            } catch (final IOException ex) {
+                response.setStatusCode(500);
+            }
+        };
+    }
+
+    private static void awaitForCondition(Supplier<Boolean> condition) throws InterruptedException {
```

Comment:
There is only 1 place when we need that, due to the fact that `failureRequestIndexer` is not immediately re-adding the requests but does it on the next batch interval only